### ISSUE TITLE
(From Fuzzer) Incremental resolver: ENFORCE failure core/types/typemaps.cc:420 params.size() == targs.size()

### ISSUE THREAD

**Issue by jvilk-stripe** [sorbet/sorbet#1086]
#### Input

May be related to https://github.com/sorbet/sorbet/issues/1080. I filed this bug separately because it only triggers with `--stress-incremental-resolver`, and the stack trace is slightly different.

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0A%20%20lansi%0A%0Aclass%20Parent%20exte%20%20A%20%3DSig%0A%20%20A%20%3D%20type_member%0A%20sig%20%7Bparams(k%3AA).void%7D%0A%20%20def%20%5B%5D(k)%0A%20%20end%0Aend%0Aclass%20C%20%3C%20Parent%0A%20def%20foo(which)%20%20%20%20self%5B%22foo%22%5D%20%20end%0Aend)

```ruby
# typed: true
  lansi

class Parent exte  A =Sig
  A = type_member
 sig {params(k:A).void}
  def [](k)
  end
end
class C < Parent
 def foo(which)    self["foo"]  end
end
```

#### Observed output

Run with `--stress-incremental-resolver` to trigger:

```
fuzz_crashers/min/crash-6a19eeda875aeff2f8fd27cfa91c25719994a09f.done:5: Redefining constant Parent::A https://srb.help/4011
     5 |  A = type_member
          ^
    fuzz_crashers/min/crash-6a19eeda875aeff2f8fd27cfa91c25719994a09f.done:4: Previous definition
     4 |class Parent exte  A =Sig
                           ^

fuzz_crashers/min/crash-6a19eeda875aeff2f8fd27cfa91c25719994a09f.done:4: Unable to resolve constant Sig https://srb.help/5002
     4 |class Parent exte  A =Sig
                              ^^^
    https://github.com/sorbet/sorbet/tree/master/rbi/sorbet/t.rbi#L2: Did you mean: T::Sig?
     2 |module T::Sig
        ^^^^^^^^^^^^^

fuzz_crashers/min/crash-6a19eeda875aeff2f8fd27cfa91c25719994a09f.done:10: Type A declared by parent Parent must be re-declared in C https://srb.help/5014
    10 |class C < Parent
        ^^^^^^^^^^^^^^^^
    fuzz_crashers/min/crash-6a19eeda875aeff2f8fd27cfa91c25719994a09f.done:5: A declared in parent here
     5 |  A = type_member
          ^^^^^^^^^^^^^^^

fuzz_crashers/min/crash-6a19eeda875aeff2f8fd27cfa91c25719994a09f.done:175: Redefining constant A https://srb.help/4012
     175 |class Parent exte  A =Sig
                             ^^^^^^
    fuzz_crashers/min/crash-6a19eeda875aeff2f8fd27cfa91c25719994a09f.done:52: Previous definition
    52 |
    53 |
    54 |
    55 |
    56 |
       |...
    63 |
    64 |
    65 |
    66 |
    67 |

fuzz_crashers/min/crash-6a19eeda875aeff2f8fd27cfa91c25719994a09f.done:176: Redefining constant Parent::A https://srb.help/4011
     176 |  A = type_member
            ^
    fuzz_crashers/min/crash-6a19eeda875aeff2f8fd27cfa91c25719994a09f.done:175: Previous definition
     175 |class Parent exte  A =Sig
                             ^
Exception::raise(): core/types/typemaps.cc:420 enforced condition params.size() == targs.size() has failed: (no message provided)

Backtrace:
sorbet::Exception::enforce_handler(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int) (in librealmain.so) (Exception.h:47)
sorbet::core::LambdaParam::_instantiate(sorbet::core::Context, absl::InlinedVector<sorbet::core::SymbolRef, 4ul, std::__1::allocator<sorbet::core::SymbolRef> > const&, std::__1::vector<sorbet::core::TypePtr, std::__1::allocator<sorbet::core::TypePtr> > const&) (in libcore.so) (typemaps.cc:420)
sorbet::core::Types::instantiate(sorbet::core::Context, sorbet::core::TypePtr const&, absl::InlinedVector<sorbet::core::SymbolRef, 4ul, std::__1::allocator<sorbet::core::SymbolRef> > const&, std::__1::vector<sorbet::core::TypePtr, std::__1::allocator<sorbet::core::TypePtr> > const&) (in libcore.so) (typemaps.cc:15)
sorbet::core::Types::resultTypeAsSeenFrom(sorbet::core::Context, sorbet::core::TypePtr, sorbet::core::SymbolRef, sorbet::core::SymbolRef, std::__1::vector<sorbet::core::TypePtr, std::__1::allocator<sorbet::core::TypePtr> > const&) (in libcore.so) (types.cc:576)
sorbet::core::(anonymous namespace)::matchArgType(sorbet::core::Context, sorbet::core::TypeConstraint&, sorbet::core::Loc, sorbet::core::Loc, sorbet::core::SymbolRef, sorbet::core::SymbolRef, sorbet::core::TypeAndOrigins const&, sorbet::core::ArgInfo const&, sorbet::core::TypePtr const&, std::__1::vector<sorbet::core::TypePtr, std::__1::allocator<sorbet::core::TypePtr> >&, sorbet::core::Loc, bool) (in libcore.so) (calls.cc:136)
sorbet::core::dispatchCallSymbol(sorbet::core::Context, sorbet::core::DispatchArgs, sorbet::core::Type const*, sorbet::core::SymbolRef, std::__1::vector<sorbet::core::TypePtr, std::__1::allocator<sorbet::core::TypePtr> >&) (in libcore.so) (calls.cc:557)
sorbet::core::AppliedType::dispatchCall(sorbet::core::Context, sorbet::core::DispatchArgs) (in libcore.so) (calls.cc:820)
sorbet::infer::Environment::processBinding(sorbet::core::Context, sorbet::cfg::Binding&, int, int, sorbet::infer::KnowledgeFilter&, sorbet::core::TypeConstraint&, sorbet::core::TypePtr&)::$_12::operator()(sorbet::cfg::Send*) const (in libinfer.so) (environment.cc:769)
sorbet::infer::Environment::processBinding(sorbet::core::Context, sorbet::cfg::Binding&, int, int, sorbet::infer::KnowledgeFilter&, sorbet::core::TypeConstraint&, sorbet::core::TypePtr&) (in libinfer.so) (environment.cc:745)
sorbet::infer::Inference::run(sorbet::core::Context, std::__1::unique_ptr<sorbet::cfg::CFG, std::__1::default_delete<sorbet::cfg::CFG> >) (in libinfer.so) (inference.cc:162)
sorbet::realmain::pipeline::CFGCollectorAndTyper::preTransformMethodDef(sorbet::core::Context, std::__1::unique_ptr<sorbet::ast::MethodDef, std::__1::default_delete<sorbet::ast::MethodDef> >) (in libpipeline.so) (pipeline.cc:54)
sorbet::ast::PostPonePreTransform_MethodDef<sorbet::realmain::pipeline::CFGCollectorAndTyper, sorbet::core::Context, true>::call(sorbet::core::Context, std::__1::unique_ptr<sorbet::ast::MethodDef, std::__1::default_delete<sorbet::ast::MethodDef> >, sorbet::realmain::pipeline::CFGCollectorAndTyper&) (in libpipeline.so) (treemap.h:204)
sorbet::ast::TreeMapper<sorbet::realmain::pipeline::CFGCollectorAndTyper, sorbet::core::Context>::mapMethodDef(std::__1::unique_ptr<sorbet::ast::MethodDef, std::__1::default_delete<sorbet::ast::MethodDef> >, sorbet::core::Context) (in libpipeline.so) (treemap.h:293)
sorbet::ast::TreeMapper<sorbet::realmain::pipeline::CFGCollectorAndTyper, sorbet::core::Context>::mapIt(std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> >, sorbet::core::Context) (in libpipeline.so) (treemap.h:666)
sorbet::ast::TreeMapper<sorbet::realmain::pipeline::CFGCollectorAndTyper, sorbet::core::Context>::mapClassDef(std::__1::unique_ptr<sorbet::ast::ClassDef, std::__1::default_delete<sorbet::ast::ClassDef> >, sorbet::core::Context) (in libpipeline.so) (treemap.h:281)
sorbet::ast::TreeMapper<sorbet::realmain::pipeline::CFGCollectorAndTyper, sorbet::core::Context>::mapIt(std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> >, sorbet::core::Context) (in libpipeline.so) (treemap.h:672)
sorbet::ast::TreeMapper<sorbet::realmain::pipeline::CFGCollectorAndTyper, sorbet::core::Context>::mapInsSeq(std::__1::unique_ptr<sorbet::ast::InsSeq, std::__1::default_delete<sorbet::ast::InsSeq> >, sorbet::core::Context) (in libpipeline.so) (treemap.h:598)
sorbet::ast::TreeMapper<sorbet::realmain::pipeline::CFGCollectorAndTyper, sorbet::core::Context>::mapIt(std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> >, sorbet::core::Context) (in libpipeline.so) (treemap.h:668)
std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> > sorbet::ast::TreeMap::apply<sorbet::core::Context, sorbet::realmain::pipeline::CFGCollectorAndTyper>(sorbet::core::Context, sorbet::realmain::pipeline::CFGCollectorAndTyper&, std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> >) (in libpipeline.so) (treemap.h:712)
sorbet::realmain::pipeline::typecheckOne(sorbet::core::Context, sorbet::ast::ParsedFile, sorbet::realmain::options::Options const&) (in libpipeline.so) (pipeline.cc:693)
sorbet::realmain::pipeline::typecheck(std::__1::unique_ptr<sorbet::core::GlobalState, std::__1::default_delete<sorbet::core::GlobalState> >&, std::__1::vector<sorbet::ast::ParsedFile, std::__1::allocator<sorbet::ast::ParsedFile> >, sorbet::realmain::options::Options const&, sorbet::WorkerPool&)::$_3::operator()() const (in libpipeline.so) (pipeline.cc:938)
decltype(std::__1::forward<sorbet::realmain::pipeline::typecheck(std::__1::unique_ptr<sorbet::core::GlobalState, std::__1::default_delete<sorbet::core::GlobalState> >&, std::__1::vector<sorbet::ast::ParsedFile, std::__1::allocator<sorbet::ast::ParsedFile> >, sorbet::realmain::options::Options const&, sorbet::WorkerPool&)::$_3&>(fp)()) std::__1::__invoke<sorbet::realmain::pipeline::typecheck(std::__1::unique_ptr<sorbet::core::GlobalState, std::__1::default_delete<sorbet::core::GlobalState> >&, std::__1::vector<sorbet::ast::ParsedFile, std::__1::allocator<sorbet::ast::ParsedFile> >, sorbet::realmain::options::Options const&, sorbet::WorkerPool&)::$_3&>(sorbet::realmain::pipeline::typecheck(std::__1::unique_ptr<sorbet::core::GlobalState, std::__1::default_delete<sorbet::core::GlobalState> >&, std::__1::vector<sorbet::ast::ParsedFile, std::__1::allocator<sorbet::ast::ParsedFile> >, sorbet::realmain::options::Options const&, sorbet::WorkerPool&)::$_3&&&) (in libpipeline.so) (type_traits:4353)
sorbet::WorkerPoolImpl::multiplexJob(std::__1::basic_string_view<char, std::__1::char_traits<char> >, std::__1::function<void ()>) (in libconcurrency.so) (WorkerPoolImpl.cc:0)
sorbet::realmain::pipeline::typecheck(std::__1::unique_ptr<sorbet::core::GlobalState, std::__1::default_delete<sorbet::core::GlobalState> >&, std::__1::vector<sorbet::ast::ParsedFile, std::__1::allocator<sorbet::ast::ParsedFile> >, sorbet::realmain::options::Options const&, sorbet::WorkerPool&) (in libpipeline.so) (pipeline.cc:927)
sorbet::realmain::realmain(int, char**) (in librealmain.so) (realmain.cc:427)
main (in sorbet) (main.cc:7)
start (in libdyld.dylib) + 1
0x00000003 (in sorbet)


fuzz_crashers/min/crash-6a19eeda875aeff2f8fd27cfa91c25719994a09f.done:182: Failed to type (backtrace is above) https://srb.help/1001
     182 | def foo(which)    self["foo"]  end
                             ^^^^^^^^^^^

fuzz_crashers/min/crash-6a19eeda875aeff2f8fd27cfa91c25719994a09f.done:182: Failed to process tree (backtrace is above) https://srb.help/1001
     182 | def foo(which)    self["foo"]  end
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

fuzz_crashers/min/crash-6a19eeda875aeff2f8fd27cfa91c25719994a09f.done:175: Unable to resolve constant Sig https://srb.help/5002
     175 |class Parent exte  A =Sig
                                ^^^
    https://github.com/sorbet/sorbet/tree/master/rbi/sorbet/t.rbi#L2: Did you mean: T::Sig?
     2 |module T::Sig
        ^^^^^^^^^^^^^

fuzz_crashers/min/crash-6a19eeda875aeff2f8fd27cfa91c25719994a09f.done:173: Method lansi does not exist on T.class_of(<root>) https://srb.help/7003
     173 |  lansi
            ^^^^^

fuzz_crashers/min/crash-6a19eeda875aeff2f8fd27cfa91c25719994a09f.done:175: Method exte does not exist on T.class_of(Parent) https://srb.help/7003
     175 |class Parent exte  A =Sig
                       ^^^^^^^^^^^^
    https://github.com/sorbet/sorbet/tree/master/rbi/core/kernel.rbi#L189: Did you mean: Kernel#extend?
     189 |  def extend(mod); end
            ^^^^^^^^^^^^^^^

fuzz_crashers/min/crash-6a19eeda875aeff2f8fd27cfa91c25719994a09f.done:177: Method sig does not exist on T.class_of(Parent) https://srb.help/7003
     177 | sig {params(k:A).void}
           ^^^^^^^^^^^^^^^^^^^^^^

fuzz_crashers/min/crash-6a19eeda875aeff2f8fd27cfa91c25719994a09f.done:177: Method params does not exist on T.class_of(Parent) https://srb.help/7003
     177 | sig {params(k:A).void}
                ^^^^^^^^^^^

fuzz_crashers/min/crash-6a19eeda875aeff2f8fd27cfa91c25719994a09f.done: Exception in cfg+infer: fuzz_crashers/min/crash-6a19eeda875aeff2f8fd27cfa91c25719994a09f.done (backtrace is above) https://srb.help/1001
Errors: 13
```

#### Expected behavior

The ENFORCE does not fail.

**Comment by aisamanra** [comment]
This appears to have been fixed as well with the #1080 problem, but I'll wait for confirmation before closing it.

**Comment by aisamanra** [comment]
I've re-tested this a few times and it's gone. I'm comfortable closing it at this point.
